boot/assets_test.go

```diff
@@ -0,0 +1,305 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot_test
+
+import (
+	"crypto"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/testutil"
+)
+
+type assetsSuite struct {
+	baseBootenvSuite
+}
+
+var _ = Suite(&assetsSuite{})
+
+func checkContentGlob(c *C, glob string, expected []string) {
+	l, err := filepath.Glob(glob)
+	c.Assert(err, IsNil)
+	c.Check(l, DeepEquals, expected)
+}
+
+func (s *assetsSuite) TestAssetsCacheAddDrop(c *C) {
+	cacheDir := c.MkDir()
+	d := c.MkDir()
+
+	cache := boot.NewTrustedAssetsCache(cacheDir)
+
+	data := []byte("foobar")
+	// SHA3-384
+	hash := "0fa8abfbdaf924ad307b74dd2ed183b9a4a398891a2f6bac8fd2db7041b77f068580f9c6c66f699b496c2da1cbcc7ed8"
+	err := ioutil.WriteFile(filepath.Join(d, "foobar"), data, 0644)
+	c.Assert(err, IsNil)
+
+	// add a new file
+	ta, err := cache.Add(filepath.Join(d, "foobar"), "grub", "grubx64.efi")
+	c.Assert(err, IsNil)
+	c.Check(filepath.Join(cacheDir, "grub", fmt.Sprintf("grubx64.efi-%s", hash)), testutil.FileEquals, string(data))
+	c.Check(ta, NotNil)
+
+	// try the same file again
+	taAgain, err := cache.Add(filepath.Join(d, "foobar"), "grub", "grubx64.efi")
+	c.Assert(err, IsNil)
+	// file already cached
+	c.Check(filepath.Join(cacheDir, "grub", fmt.Sprintf("grubx64.efi-%s", hash)), testutil.FileEquals, string(data))
+	// and there's just one entry in the cache
+	checkContentGlob(c, filepath.Join(cacheDir, "grub", "*"), []string{
+		filepath.Join(cacheDir, "grub", fmt.Sprintf("grubx64.efi-%s", hash)),
+	})
+	// let go-check do the deep equals check
+	c.Check(taAgain, DeepEquals, ta)
+
+	// same data but different asset name
+	taDifferentAsset, err := cache.Add(filepath.Join(d, "foobar"), "grub", "bootx64.efi")
+	c.Assert(err, IsNil)
+	// new entry in cache
+	c.Check(filepath.Join(cacheDir, "grub", fmt.Sprintf("bootx64.efi-%s", hash)), testutil.FileEquals, string(data))
+	// 2 files now
+	checkContentGlob(c, filepath.Join(cacheDir, "grub", "*"), []string{
+		filepath.Join(cacheDir, "grub", fmt.Sprintf("bootx64.efi-%s", hash)),
+		filepath.Join(cacheDir, "grub", fmt.Sprintf("grubx64.efi-%s", hash)),
+	})
+	c.Check(taDifferentAsset, NotNil)
+
+	// same source, data (new hash), existing asset name
+	newData := []byte("new foobar")
+	newH := crypto.SHA256.New()
```

Comment:
this is wrong now, no? or is the test wrong?

Reply:
No, just forgot to remove this bit.